### PR TITLE
Reapply "Upgrade to Curator 4"

### PR DESCRIPTION
**Pull request by hmusum**
Reverts vespa-engine/vespa#15904

Please review only. Need to fix some code related to deletion of tenants before this can go in.

### CODE REVIEW

File: zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java [hunk 1]
```diff
@@ -776,6 +875,11 @@ public CreateBackgroundModeACLable compressed() {
             throw new UnsupportedOperationException("Not implemented in MockCurator");
         }
 
+        @Override
+        public ACLCreateModeStatBackgroundPathAndBytesable<String> withProtection() {
+            return null;
```
Comment (hakonhall) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java:880]: return this or throw?

File: zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java [hunk 2]
```diff
@@ -813,6 +917,47 @@ public String forPath(String s, byte[] bytes) throws Exception {
         public ErrorListenerPathAndBytesable<String> inBackground(BackgroundCallback backgroundCallback, Object o, Executor executor) {
             throw new UnsupportedOperationException("Not implemented in MockCurator");
         }
+
+        @Override
+        public CreateBuilderMain withTtl(long l) {
+            return null;
```
Comment (hakonhall) [inline zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java:923]: These should throw or return this?

**Review by hakonhall — APPROVED** [review]

File: zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java (hunk 2 above)
Comment (hmusum) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java]: Fixed some of these

File: zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java (hunk 1 above)
Comment (hmusum) [reply zkfacade/src/main/java/com/yahoo/vespa/curator/mock/MockCuratorFramework.java]: Fixed

**Pull request merged**